Generated index.js doesn't work in Windows machine

## Step to reproduce
1. On a Windows machine, clone beverage-buddy starter
2. Run `mvn`
## Actual behavior
Frontend compilation error
```
[webpack] ERROR dev-webpack - ERROR in ../target
[webpack] ERROR dev-webpack - Module not found: Error: Can't resolve './frontendgenerated-flow-imports.js' in 'C:\WorkFolder\beverage-starter-flow\target'
[webpack] ERROR dev-webpack -  @ ../target 5:17-63
[webpack] ERROR dev-webpack -
[webpack] ERROR dev-webpack - ERROR in ../target
[webpack] ERROR dev-webpack - Module not found: Error: Can't resolve './frontendgenerated-flow-imports.js' in 'C:\WorkFolder\beverage-starter-flow\target'
```
The generated `index.js` in target folder uses a wrong path to `generated-flow-import.js`
```
const flow = new Flow({
  imports: () => import('./frontend\generated-flow-imports.js')
});
```
## Expect behavior
It should start the application without errors

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -19,6 +19,7 @@ import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
@@ -600,6 +601,31 @@ public class FrontendUtils {
         return file.exists()
                 && FileUtils.readFileToString(file, StandardCharsets.UTF_8)
                         .contains("./webpack.generated.js");
+    }
+
+    /**
+     * Get relative path from a source path to a target path in Unix form. All
+     * the Windows' path separator will be replaced.
+     * 
+     * @param source
+     *            the source path
+     * @param target
+     *            the target path
+     * @return unix relative path from source to target
+     */
+    public static String getUnixRelativePath(Path source, Path target) {
+        return getUnixPath(source.relativize(target));
+    }
+
+    /**
+     * Get path as a String in Unix form.
+     * 
+     * @param source
+     *            path to get
+     * @return path as a String in Unix form.
+     */
+    public static String getUnixPath(Path source) {
+        return source.toString().replaceAll("\\\\", "/");
     }
 
     /**

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/NodeUpdater.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/NodeUpdater.java
@@ -221,11 +221,11 @@ public abstract class NodeUpdater implements FallibleCommand {
         // dependency for the custom package.json placed in the generated
         // folder.
         try {
-            String customPkg = "./" + npmFolder.getAbsoluteFile().toPath()
-                    .relativize(generatedFolder.getAbsoluteFile().toPath())
-                    .toString();
+            String customPkg = "./" + FrontendUtils.getUnixRelativePath(
+                    npmFolder.getAbsoluteFile().toPath(),
+                    generatedFolder.getAbsoluteFile().toPath());
             added = addDependency(packageJson, DEPENDENCIES, DEP_NAME_FLOW_DEPS,
-                    customPkg.replaceAll("\\\\", "/")) || added;
+                    customPkg) || added;
         } catch (IllegalArgumentException iae) {
             log().error("Exception in relativization of '{}' to '{}'",
                     npmFolder.getAbsoluteFile().toPath(),

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskGenerateIndexJs.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskGenerateIndexJs.java
@@ -70,8 +70,9 @@ public class TaskGenerateIndexJs extends AbstractTaskClientGenerator {
     protected String getFileContent() throws IOException {
         String indexTemplate = IOUtils
                 .toString(getClass().getResourceAsStream(INDEX_JS), UTF_8);
-        String relativizedImport = ensureValidRelativePath(outputDirectory
-                .toPath().relativize(generatedImports.toPath()).toString());
+        String relativizedImport = ensureValidRelativePath(
+                FrontendUtils.getUnixRelativePath(outputDirectory.toPath(),
+                        generatedImports.toPath()));
         return indexTemplate.replace("[to-be-generated-by-flow]",
                 relativizedImport);
     }
@@ -85,11 +86,10 @@ public class TaskGenerateIndexJs extends AbstractTaskClientGenerator {
      * @return valid import path
      */
     static String ensureValidRelativePath(String relativePath) {
-        String relativizedImport = relativePath.replaceAll("\\\\", "/");
-        if (!relativizedImport.startsWith(".")) {
-            relativizedImport = "./" + relativizedImport;
+        if (!relativePath.startsWith(".")) {
+            relativePath = "./" + relativePath;
         }
-        return relativizedImport;
+        return relativePath;
     }
 
 }

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskGenerateIndexJs.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskGenerateIndexJs.java
@@ -18,11 +18,10 @@ package com.vaadin.flow.server.frontend;
 import java.io.File;
 import java.io.IOException;
 
-import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_JS;
-import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_TS;
 import org.apache.commons.io.IOUtils;
 
-import static com.vaadin.flow.server.frontend.FrontendUtils.IMPORTS_NAME;
+import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_JS;
+import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_TS;
 import static java.nio.charset.StandardCharsets.UTF_8;
 
 /**
@@ -71,9 +70,26 @@ public class TaskGenerateIndexJs extends AbstractTaskClientGenerator {
     protected String getFileContent() throws IOException {
         String indexTemplate = IOUtils
                 .toString(getClass().getResourceAsStream(INDEX_JS), UTF_8);
-        String relativizedImport = outputDirectory.toPath()
-                .relativize(generatedImports.toPath()).toFile().getPath();
+        String relativizedImport = ensureValidRelativePath(outputDirectory
+                .toPath().relativize(generatedImports.toPath()).toString());
         return indexTemplate.replace("[to-be-generated-by-flow]",
-                "./" + relativizedImport);
+                relativizedImport);
     }
+
+    /**
+     * Ensure that the given relative path is valid as an import path. NOTE:
+     * expose only for testing purpose.
+     * 
+     * @param relativePath
+     *            given relative path
+     * @return valid import path
+     */
+    static String ensureValidRelativePath(String relativePath) {
+        String relativizedImport = relativePath.replaceAll("\\\\", "/");
+        if (!relativizedImport.startsWith(".")) {
+            relativizedImport = "./" + relativizedImport;
+        }
+        return relativizedImport;
+    }
+
 }

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskUpdateWebpack.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskUpdateWebpack.java
@@ -24,16 +24,16 @@ import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.List;
 
-import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_HTML;
-import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_JS;
-import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_TS;
-import static com.vaadin.flow.server.frontend.FrontendUtils.TARGET;
 import org.apache.commons.io.FileUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import com.vaadin.flow.server.FallibleCommand;
 
+import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_HTML;
+import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_JS;
+import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_TS;
+import static com.vaadin.flow.server.frontend.FrontendUtils.TARGET;
 import static com.vaadin.flow.server.frontend.FrontendUtils.WEBPACK_CONFIG;
 import static com.vaadin.flow.server.frontend.FrontendUtils.WEBPACK_GENERATED;
 
@@ -209,10 +209,11 @@ public class TaskUpdateWebpack implements FallibleCommand {
     }
 
     private String getEscapedRelativeWebpackPath(Path path) {
-        Path relativePath = path.isAbsolute()
-                ? webpackConfigPath.relativize(path)
-                : path;
-        return relativePath.toString().replaceAll("\\\\", "/");
+        if (path.isAbsolute()) {
+            return FrontendUtils.getUnixRelativePath(webpackConfigPath, path);
+        } else {
+            return FrontendUtils.getUnixPath(path);
+        }
     }
 
     private Logger log() {

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -20,6 +20,7 @@ import java.io.File;
 import java.io.IOException;
 import java.io.PrintStream;
 import java.io.UnsupportedEncodingException;
+import java.nio.file.Path;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
@@ -29,6 +30,7 @@ import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
 import org.junit.rules.TemporaryFolder;
+import org.mockito.Mockito;
 import org.slf4j.LoggerFactory;
 
 import com.vaadin.flow.server.frontend.FrontendUtils.UnknownVersionException;
@@ -214,6 +216,29 @@ public class FrontendUtilsTest {
     @Test(expected = IOException.class)
     public void parseEmptyToolVersions() throws IOException {
         FrontendUtils.parseVersion(" \n");
+    }
+
+    @Test
+    public void should_getUnixRelativePath_when_givenTwoPaths() {
+        Path sourcePath = Mockito.mock(Path.class);
+        Path relativePath = Mockito.mock(Path.class);
+        Mockito.when(sourcePath.relativize(Mockito.any()))
+                .thenReturn(relativePath);
+        Mockito.when(relativePath.toString())
+                .thenReturn("this\\is\\windows\\path");
+
+        String relativeUnixPath = FrontendUtils.getUnixRelativePath(sourcePath,
+                tmpDir.getRoot().toPath());
+        Assert.assertEquals(
+                "Should replace windows path separator with unix path separator",
+                "this/is/windows/path", relativeUnixPath);
+        Mockito.when(relativePath.toString()).thenReturn("this/is/unix/path");
+
+        relativeUnixPath = FrontendUtils.getUnixRelativePath(sourcePath,
+                tmpDir.getRoot().toPath());
+        Assert.assertEquals(
+                "Should keep the same path when it uses unix path separator",
+                "this/is/unix/path", relativeUnixPath);
     }
 
 }

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskGenerateIndexJsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskGenerateIndexJsTest.java
@@ -75,23 +75,19 @@ public class TaskGenerateIndexJsTest {
                 IOUtils.toString(taskGenerateIndexJs.getGeneratedFile().toURI(),
                         StandardCharsets.UTF_8));
     }
-@Test
-    public void should_notContainWindowSeparator_whenRelativizingPath() throws Exception
-    {
-        String windowSeparator = TaskGenerateIndexJs
-                .ensureValidRelativePath("frontend\\generated-flow-imports.js");
-        Assert.assertEquals("Window separator should be replaced",
-                "./frontend/generated-flow-imports.js", windowSeparator);
 
-        String unixSeparator = TaskGenerateIndexJs
-                .ensureValidRelativePath("frontend/generated-flow-imports.js");
-        Assert.assertEquals("Unix separator should be kept",
-                "./frontend/generated-flow-imports.js", unixSeparator);
-
+    @Test
+    public void should_ensureValidRelativePath_whenItHasNoRelativePrefix() {
         String customPath = TaskGenerateIndexJs.ensureValidRelativePath(
                 "../custom-frontend/generated-flow-imports.js");
         Assert.assertEquals(
                 "Should not append './' if it is already a relative path",
                 "../custom-frontend/generated-flow-imports.js", customPath);
+
+        customPath = TaskGenerateIndexJs.ensureValidRelativePath(
+                "custom-frontend/generated-flow-imports.js");
+        Assert.assertEquals(
+                "Should append './' if it doesn't start with a relative path",
+                "./custom-frontend/generated-flow-imports.js", customPath);
     }
 }

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskGenerateIndexJsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskGenerateIndexJsTest.java
@@ -20,13 +20,14 @@ import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 
-import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_JS;
 import org.apache.commons.io.IOUtils;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
+
+import static com.vaadin.flow.server.frontend.FrontendUtils.INDEX_JS;
 
 public class TaskGenerateIndexJsTest {
     @Rule
@@ -59,7 +60,7 @@ public class TaskGenerateIndexJsTest {
     }
 
     @Test
-    public void should_generateIndexHtml_IndexJsNotExist() throws Exception {
+    public void should_generateIndexJs_IndexJsNotExist() throws Exception {
 
         taskGenerateIndexJs.execute();
         Assert.assertTrue(
@@ -73,5 +74,24 @@ public class TaskGenerateIndexJsTest {
                 taskGenerateIndexJs.getFileContent(),
                 IOUtils.toString(taskGenerateIndexJs.getGeneratedFile().toURI(),
                         StandardCharsets.UTF_8));
+    }
+@Test
+    public void should_notContainWindowSeparator_whenRelativizingPath() throws Exception
+    {
+        String windowSeparator = TaskGenerateIndexJs
+                .ensureValidRelativePath("frontend\\generated-flow-imports.js");
+        Assert.assertEquals("Window separator should be replaced",
+                "./frontend/generated-flow-imports.js", windowSeparator);
+
+        String unixSeparator = TaskGenerateIndexJs
+                .ensureValidRelativePath("frontend/generated-flow-imports.js");
+        Assert.assertEquals("Unix separator should be kept",
+                "./frontend/generated-flow-imports.js", unixSeparator);
+
+        String customPath = TaskGenerateIndexJs.ensureValidRelativePath(
+                "../custom-frontend/generated-flow-imports.js");
+        Assert.assertEquals(
+                "Should not append './' if it is already a relative path",
+                "../custom-frontend/generated-flow-imports.js", customPath);
     }
 }

